### ISSUE TITLE
Inquiry about Visualization of Diffusion Model's Prediction Results

### ISSUE THREAD

**Issue by iris0329** [Haochen-Wang409/ross3d#14]
Dear Author,

I wonder if there are any visualizations available for the image prediction effects of the trained diffusion model. I have used the **released ckpt** and called the **sample() function** provided in this repo, but so far, I can only visualize noise instead of the expected predicted images.

Could you please provide some guidance on this matter? Your assistance would be greatly appreciated. Thank you for your time and help.

**Comment by Haochen-Wang409** [comment]
At this time, we do not plan to release this part. However, there is an implementation here: https://github.com/Haochen-Wang409/ross/issues/14. You may refer to that.

**Comment by iris0329** [comment]
Hi @Haochen-Wang409 ，
Thank you for your guidance.

Following the approach you mentioned, I attempted a simple reconstruction on the training set to verify the process: I used a batch of 32 frames of data, taking 24 of them as conditions and trying to reconstruct the remaining 8 frames (targets)

However, the visualized results appear as noise. I’ve attached a screenshot of these results for your reference. I would like to ask: does this visualization result meet the expected outcome?

# input frames:
<img width="1600" height="800" alt="Image" src="https://github.com/user-attachments/assets/a923d540-17fd-4579-8780-a524324b059e" />

# reconstruction result:
<img width="1810" height="228" alt="Image" src="https://github.com/user-attachments/assets/1ed9aea1-6320-4015-b2aa-01652acb292f" />


I’ve also attached the current core code snippet for your review. 

```python
with torch.amp.autocast('cuda', dtype=torch.float32):
    h = w = int(image_hidden_states.shape[1] ** 0.5)
    image_hidden_states_ln = self.model.mm_inv_projector.ln_pre(image_hidden_states)
    image_hidden_states_ln = rearrange(image_hidden_states_ln, 'b (h w) c -> b c h w', h=h, w=w).contiguous() # [B=24, dim=3584, h=14, w=14]

    sampled_latents = self.model.mm_inv_projector.sample(z=image_hidden_states_ln.float(), hw=h, temperature=1.0, cfg=1.0, bev=False, gen_bs=8) # [B=8, dim=64, h=14, w=14]

    sampled_latents = rearrange(sampled_latents, 'b (c p1 p2) h w -> b c (h p1) (w p2)', c=sampled_latents.shape[1]//4, p1=2, p2=2).contiguous() 
    sampled_latents = sampled_latents / self.model.mm_pixel_decoder.scaling_factor + self.model.mm_pixel_decoder.shift_factor
    generated_images = self.model.mm_pixel_decoder.decode(sampled_latents)['sample'] 

    gens_01 = (generated_images * 0.5 + 0.5).clamp(0, 1)
    vutils.save_image(gens_01, 'recon_grid.png', nrow=8, padding=2)
```

The main difference between my implementation and the snippet in the ROSS issue lies in the handling of the batch size for the condition and target (but same as the training setting):

- Batch size for the condition: I provided 24 (unmasked frames).
- Batch size for the target: I am trying to predict 8 (masked frames).

Thank you again for your time and help.

**Comment by Haochen-Wang409** [comment]
The batch size for the condition should be 32. This is because we only mask RGB frames, but take depth maps and camera poses for masked frames as inputs.

The reconstruction results are expected to be unsatisfactory, as Ross further trains the denoiser on ImageNet-1K to obtain reasonable reconstruction results, while we did not do that.

**Comment by iris0329** [comment]
Hi Haochen,

Thank you for your prompt reply.  

I’m still a bit confused about your reply: "This is because we only mask RGB frames, but take depth maps and camera poses for masked frames as inputs." I haven’t come across any operations related to depth maps or camera poses in the code, so I’m wondering if I might have missed something crucial here?  

Additionally, if the batch size for the condition should be 32 (with 8 being targets), this seems to differ from the training setup, where it was 24 conditions and 8 targets. I’m curious about why this adjustment is needed during inference.  

Thank you again for your patience and assistance.  

Best regards,  
iris0329

**Comment by Haochen-Wang409** [comment]
1. Depth maps and camera poses are used to encode the ```world_position_embedding``` [here](https://github.com/Haochen-Wang409/ross3d/blob/main/ross3d/model/ross3d_arch.py#L699).
2. Let's elaborate how ```image_hidden_states``` was obtained step by step:
    - First, in the function [here](https://github.com/Haochen-Wang409/ross3d/blob/main/ross3d/model/ross3d_arch.py#L440), we obtain ```boi_ids``` and ```eoi_ids``` with a length of ```num_frames```, i.e., 32.
    - Next, in the function [here](https://github.com/Haochen-Wang409/ross3d/blob/main/ross3d/model/ross3d_arch.py#L1089), ```image_hidden_states``` are obtained with *each* frame.

Therefore, ```image_hidden_states``` should be 32 instead of 24.

Still, the reconstruction results are expected to be unsatisfactory (similar to noise), as the original Ross further trains the denoiser on ImageNet-1K to obtain reasonable reconstruction results, while we did not do that.

**Comment by iris0329** [comment]
Hi @Haochen-Wang409,

Thank you for your reply. I understand that the expected results are poor due to the missing ImageNet fine-tuning.

However, I still have some confusion regarding the input structure. My question is specifically about predicting the masked video frames, not the BEV map prediction process.

Referring to the code:
In the [training code here](https://github.com/Haochen-Wang409/ross3d/blob/main/ross3d/model/ross3d_arch.py#L1107), when `mask.sum() > 0`, **24 frames are kept** and used as conditions.
Then [here at L1130](https://github.com/Haochen-Wang409/ross3d/blob/main/ross3d/model/ross3d_arch.py#L1130), the condition is the 24 frames (image_hidden_states encoded by the LLM), and the prediction target is the VAE latents of the randomly masked 8 frames.

This is the diffusion model's **training** process. Shouldn't the **evaluation** (inference) process be consistent with the training?

If the condition during inference is the full 32 frames, what is the number of the target then? 

Thank you again for your patience and invaluable help.

Best regards,
iris0329

**Comment by Haochen-Wang409** [comment]
Yes, you are right. What I want to emphasize is that these 24 ```image_hidden_states``` are actually interacting with the remaining 8.

**Comment by iris0329** [comment]
Ah I see now. thanks a lot.

获取 Outlook for iOS<https://aka.ms/o0ukef>
________________________________
发件人: Haochen Wang ***@***.***>
发送时间: Sunday, August 24, 2025 7:21:02 PM
收件人: Haochen-Wang409/ross3d ***@***.***>
抄送: Rong LI ***@***.***>; Author ***@***.***>
主题: Re: [Haochen-Wang409/ross3d] Inquiry about Visualization of Diffusion Model's Prediction Results (Issue #14)

[https://avatars.githubusercontent.com/u/62472073?s=20&v=4]Haochen-Wang409 left a comment (Haochen-Wang409/ross3d#14)<https://github.com/Haochen-Wang409/ross3d/issues/14#issuecomment-3218030614>

Yes, you are right. What I want to emphasize is that these 24 image_hidden_states are actually interacting with the remaining 8.

―
